src/data/list/basic.lean

```diff
@@ -1911,6 +1911,17 @@ theorem mfoldr_eq_foldr (f : α → β → m β) (b l) :
   mfoldr f b l = foldr (λ a mb, mb >>= f a) (pure b) l :=
 by induction l; simp *
 
+@[simp] lemma mmap_nil {f : α → m β} : list.nil.mmap f = pure [] := rfl
+
+@[simp] lemma mmap_cons {f : α → m β} {a l} :
+  (a :: l).mmap f = f a >>= λ a', l.mmap f >>= λ l', pure (a'::l') := rfl
```

Comment:
This makes the expression a lot longer, but I suppose it does result in nice reductions for constant lists.

Reply:
I guess a question I have is whether `simp` should "run" computations. In this case, let's say `f a = none`, then the whole thing will simplify to `none`.

Reply:
What if the simplifier knows nothing about `f`? What output does that give?

Reply:
I see `data.vector2` has a corresponding simp lemma `mmap_cons`. (It's written in terms of `do` notation, which is a bit nicer to look at, but `#check` prints the type the way you've written it here so I guess they are syntactically equivalent.)

Reply:
Without knowing the use case I find it really hard to guess whether this is a good simp lemma or not. @pechersky what are you working toward with these list PRs? I'm not complaining, they're good additions regardless, but knowing the target can help with reviewing.

Reply:
I am working on proving statements about how parsers and parser combinators work. For example, in a repo I have the following working:

```
#eval run_string (mmap piece.parser' [king, queen, pawn]) "KQP" -- sum.inr [♔, ♕, ♙]
#eval run_string (mmap piece.parser' [king, queen, pawn]) "KQR" -- sum.inl "KQR\n  ^\n\nexpected: P | ♙\n"
```

and I am working on proving related lemmas. In the case of the `mmap`, it would a proposition looking like:
```
lemma mem_of_mmap {α β} {f : α → parser β} {la : list α} {lb : list β}
  (h : lb ∈ la.mmap f) : ∀ b ∈ lb, ∃ a ∈ la, b ∈ f a := sorry
```

Reply:
@digama0 @cipher1024 how do you feel about these as simp lemmas?

Reply:
I think `mmap_nil` makes sense. I can see how `mmap_cons` could make sense there's something that rubs me the wrong way about it. One observation is that it only works if `m` is a monad. You could make it more generally applicable by using `<*>` instead of `>>=`.

Reply:
`mmap` itself requires `[monad m]`. Rephrasing it in applicative syntax make the proof no longer `rfl`. I can switch it to that style if you prefer. However, most lemmas that simplify I've seen phrased in either `map` or `bind` style.

Relatedly, is there "applicative-do" support in Lean?

Reply:
I forgot about that. To change that assumption, we'd need a PR the Lean. Does it make a difference to your proofs if `mmap_cons` is not `simp`?  

Also, since both of these lemmas are provable by `simp`, you could get the same result by labelling the definition of `mmap` with `simp`.

And Lean does not have Haskell's `ApplicativeDo`. It might be implementable for Lean 4, the syntax is very flexible.

Reply:
If `mmap_cons` isn't `simp`, that's OK because I'll just remember what the name is and include it within any `simp [...]` call. Do you think it's necessary to update Lean itself, generalizing `mmap`? Should tagging `mmap` with `@[simp]` be done upstream in Lean, in this PR, or otherwise?

Reply:
@cipher1024 how do you suggest we proceed here?

Reply:
Let's replace the four lemmas ( `mmap'?_(nil|cons)` ) with `attribute [simp] mmap mmap'`